app/models/labware_creators/common_file_handling/csv_file_for_tube_rack.rb

```diff
@@ -0,0 +1,127 @@
+# frozen_string_literal: true
+
+# Part of the Labware creator classes
+module LabwareCreators
+  #
+  # This is an abstract class for handling tube rack csv files.
+  #
+  # Takes the user uploaded tube rack scan csv file, validates the content and extracts the information.
+  #
+  # This version of the rack scan file contains 2 columns, the first is the tube location (coordinate within rack)
+  # and the second is the tube barcode.
+  #
+  # Example of file content (NB. no header line):
+  # A1,FX05653780
+  # A2,NO READ
+  # etc.
+  #
+  class CommonFileHandling::CsvFileForTubeRack < CommonFileHandling::CsvFileBase
+    validates_nested :tube_rack_scan, if: :correctly_parsed?
+    validate :check_no_duplicate_rack_positions, if: :correctly_parsed?
+    validate :check_no_duplicate_tube_barcodes, if: :correctly_parsed?
```

Comment:
The method in the if condition adds an error as well as returning a boolean. Does it cause adding the same error three times as it has to be called to find out whether the validations here should be done.

Reply:
refactored